### ISSUE TITLE
Expected logic for caching field values

### ISSUE THREAD

**Issue by salsakran** [metabase/metabase#5381]
This is a catch-all issue to discuss the logic for when and how to cache field values in the sync process. This is relevant to #5363 and any subsequent PRs

Field Values are cached when-
* A database is added
* Nightly (or when otherwise scheduled once #5363 is merged)
* A database "sync" is manually triggered by a user

Current problems:
1. Sometimes we identify as low cardinality fields that don't really make sense are being cached (eg, IPs etc). This often occurs in small databases where there are < 300 rows for some tables and hence everything is low-cardinality. 
2. Sometimes users need autocomplete for a specific field that has many values. We've had issues with the 30 and now 300 limit for cached values. 
3. Sometimes a table has been regenerated, and there are old field values that no longer apply in the cache
4. Sometimes there are entries in the cached field values that correspond to bad data, and shouldn't be in the set of filter choices. 
5. With the new sampling logic, it's likely we'll miss recently added field values that occur in later rows. 

Proposed changes
1. I *think* that we should ignore the following types when caching field values
```
:type/Float
:type/URL
:type/Email
:type/Description
:type/TextLike
:type/PK
:type/FK
```

2. Add a setting to each field which is "cached values" which is automatically set using our logic above, but can be modified by the admin. This would be a way to clean up 1, 2.
3. Provide a "re-scan for cached values" which drops all current values and does a full table scan for new valued. This would be a manual fix for problems 3 + 4 + 5 above. 

Anyone have any ideas on ways to detect if problem 5? @sbelak @camsaul @thearthur ?

**Comment by sbelak** [comment]
1) we should be thresholding both cardinality/count and absolute cardinality.

5) In updates we could only sample from last row count onwards and then do a proportionate merge if we want to retain the distribution, else just fish out uniques (ideally we'd do both store the distribution and all uniques). This assumes a) default ordering is stable and b) unseen values are much more likely be inserted rather than updated. For very low cardinality fields that we know are indexed it might even make sense to do a `not in`.

**Comment by salsakran** [comment]
I just realized we're missing the simplest and possibly strongest signal here -

We should prioritize the caching of field values for any field that's used in a field-based filter clause or a dashboard parameter. If we do this on the creation of that sql param or dashboard, it should be ready when a user wants to actually use it.

**Comment by camsaul** [comment]
makes sense

**Comment by salsakran** [comment]
also, as part of the new dimensions work @kdoh + @senior started on, we really should just put explicit "cache values" + a "rescan values" knobs. 

As much as I want to get around it, I don't think there's an alternative for very large OLAP databases.  If it's tucked away it shouldn't be too in people's faces.

**Comment by camsaul** [comment]
Implemented by #5830 and #5883
